File: src/csharp/Problems/TwoSum.cs
//https://leetcode.com/problems/two-sum/

namespace LeetCode.Problems;

public sealed class TwoSum : ProblemBase
{
    public static void Run()
    {
        //var d = Run(new [] {2,7,11,15}, 9);
        //var d = Run(new [] {3,2,4}, 6);
        var d = Run(new[] { 1, 1, 1, 1, 1, 4, 1, 1, 1, 1, 1, 7, 1, 1, 1, 1, 1 }, 11);
    }

    //Option 2
    private static int[] Run(int[] nums, int target)
    {
        var map = new Dictionary<int, int>();
        for (var i = 0; i < nums.Length; i++)
        {
            if (map.TryGetValue(target - nums[i], out var index))
            {
                return new[] { index, i };
            }
            else
            {
                map.TryAdd(nums[i], i);
            }
        }

        return Array.Empty<int>();
    }

    //Option 1
    private static int[] Run1(int[] nums, int target)
    {
        for (var i = 0; i < nums.Length; i++)
        {
            for (var j = i + 1; j < nums.Length; j++)
            {
                if (nums[i] + nums[j] == target)
                {
                    return new[] { i, j };
                }
            }
        }

        return Array.Empty<int>();
    }
}

File: src/python/problems/TwoSum.py
#https://leetcode.com/problems/two-sum/

from core.ProblemBase import *

class TwoSum(ProblemBase):
    def Solution(self, nums: List[int], target: int) -> List[int]:
        map = {}

        for i in range(len(nums)):
            if target - nums[i] in map:
                return [map[target - nums[i]], i]
            map[nums[i]] = i
            
        return []

        
if __name__ == '__main__':
    TestGen(TwoSum) \
        .Add(lambda tc: tc.Param("nums", [2,7,11,15]).Param("target", 9).Result([0,1])) \
        .Add(lambda tc: tc.Param("nums", [3,2,4]).Param("target", 6).Result([1,2])) \
        .Add(lambda tc: tc.Param("nums", [3,3]).Param("target", 6).Result([0,1])) \
        .Run()
